sync2/tokens_table.go

```diff
@@ -0,0 +1,233 @@
+	// https://www.postgresql.org/docs/11/sql-select.html#SQL-DISTINCT
+	err = t.db.Select(
+		&tokens,
+		`SELECT DISTINCT ON (user_id, device_id) token_encrypted, user_id, device_id, last_seen, since
+		FROM syncv3_sync2_tokens JOIN syncv3_sync2_devices USING (user_id, device_id)
+		ORDER BY user_id, device_id, last_seen DESC
+	`)
+	if err != nil {
+		return
+	}
+	for _, token := range tokens {
+		token.AccessToken, err = t.decrypt(token.AccessTokenEncrypted)
+		if err != nil {
+			// Ignore decryption failure.
+			continue
+		}
+		token.AccessTokenHash = hashToken(token.AccessToken)
+	}
+	return
+}
+
+// Insert a new token into the table.
+func (t *TokensTable) Insert(plaintextToken, userID, deviceID string, lastSeen time.Time) (*Token, error) {
+	hashedToken := hashToken(plaintextToken)
+	encToken := t.encrypt(plaintextToken)
+	_, err := t.db.Exec(
+		`INSERT INTO syncv3_sync2_tokens(token_hash, token_encrypted, user_id, device_id, last_seen)
+		VALUES ($1, $2, $3, $4, $5)
+		ON CONFLICT (token_hash) DO NOTHING;`,
+		hashedToken, encToken, userID, deviceID, lastSeen,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return &Token{
+		AccessToken:     plaintextToken,
+		AccessTokenHash: hashedToken,
+		// Note: if this token already exists in the DB, encToken will differ from
+		// the DB token_encrypted column. (t.encrypt is nondeterministic, see e.g.
+		// https://en.wikipedia.org/wiki/Probabilistic_encryption).
+		// The rest of the program should ignore this field; it only lives here so
+		// we can Scan the DB row into the Tokens struct. Could make it private?
+		AccessTokenEncrypted: encToken,
+		UserID:               userID,
+		DeviceID:             deviceID,
+		LastSeen:             lastSeen,
+	}, nil
+}
+
+// MaybeUpdateLastSeen actions a request to update a Token struct with its last_seen value
+// in the DB. To avoid spamming the DB with a write every time a sync3 request arrives,
+// we only update the last seen timestamp or the if it is at least 24 hours old.
+// The timestamp is updated on the Token struct if and only if it is updated in the DB.
+func (t *TokensTable) MaybeUpdateLastSeen(token *Token, newLastSeen time.Time) error {
+	sinceLastSeen := newLastSeen.Sub(token.LastSeen)
+	if sinceLastSeen < (24 * time.Hour) {
+		return nil
+	}
+	_, err := t.db.Exec(
+		`UPDATE syncv3_sync2_tokens SET last_seen = $1 WHERE token_hash = $2`,
+		newLastSeen, token.AccessTokenHash,
+	)
+	if err != nil {
+		return err
+	}
+	token.LastSeen = newLastSeen
+	return nil
+}
+
+func (t *TokensTable) GetTokenAndSince(userID, deviceID, tokenHash string) (accessToken, since string, err error) {
+	var encToken, gotUserID, gotDeviceID string
+	row := t.db.QueryRow(`
+		SELECT token_encrypted, since, user_id, device_id
+		FROM syncv3_sync2_tokens JOIN syncv3_sync2_devices USING (user_id, device_id)
+		WHERE token_hash = $1;
+	`, tokenHash)
+	err = row.Scan(&encToken, &since, &gotUserID, &gotDeviceID)
```

Comment:
You'd typically inline this e.g `if err := QueryRow(...).Scan(&foo); err != nil { ... }`.